js/packages/cli/src/helpers/upload/arweave-bundle.ts

```diff
@@ -0,0 +1,288 @@
+import { readFile, stat } from 'fs/promises';
+import path from 'path';
+import Arweave from 'arweave';
+import {
+  ArweaveSigner,
+  bundleAndSignData,
+  createData,
+  DataItem,
+} from 'arbundles';
+import log from 'loglevel';
+import { EXTENSION_PNG } from '../constants';
+
+/**
+ * The Manifest object for a given asset.
+ * This object holds the contents of the asset's JSON file.
+ * Represented here in its minimal form.
+ */
+type Manifest = {
+  image: string;
+  properties: {
+    files: Array<{ type: string; uri: string }>;
+  };
+};
+
+/**
+ * The result of the processing of a set of assets file pairs, to be bundled
+ * before upload.
+ */
+type ProcessedBundleFilePairs = {
+  cacheKeys: string[];
+  dataItems: DataItem[];
+  manifestLinks: string[];
+  updatedManifests: Manifest[];
+};
+
+/**
+ * The result of the upload of a bundle, identical to ProcessedBundleFilePairs
+ * without the `dataItems` property, which holds the binary data.
+ */
+type UploadGeneratorResult = Omit<ProcessedBundleFilePairs, 'dataItems'>;
+
+// The limit for the cumulated size of filepairs to include in a single bundle.
+// arBundles has a limit of 250MB, we use our own limit way below that to
+// lower the risk for having to re-upload filepairs if the matching manifests
+// upload fail on voluminous collections.
+// Change at your own risk.
+const BUNDLE_SIZE_LIMIT = 200 * 1000 * 1000;
+
+/**
+ * Tags to include with every individual transaction.
+ */
+const BASE_TAGS = [{ name: 'App-Name', value: 'Metaplex Candy Machine' }];
+
+const CONTENT_TYPES = {
+  png: 'image/png',
+};
+
+const contentTypeTags = {
+  png: { name: 'Content-Type', value: CONTENT_TYPES['png'] },
+  json: { name: 'Content-Type', value: 'application/json' },
+};
+
+/**
+ * Create an Arweave instance with sane defaults.
+ */
+function getArweave(): Arweave {
+  return new Arweave({
+    host: 'arweave.net',
+    port: 443,
+    protocol: 'https',
+    timeout: 20000,
+    logging: false,
+    logger: console.log,
+  });
+}
+
+/**
+ * Simplistic helper to convert a bytes value to its MB counterpart.
+ */
+function sizeMB(bytes: number): number {
+  return bytes / (1000 * 1000);
```

Comment:
It's a habit of using the decimal definition, as opposed to kibibytes / mebibytes, but no strong opinion, also felt safer since [Arweave claims](https://github.com/Bundlr-Network/arbundles/blob/master/README.md#arweave-bundles):

> Gateways currently only index bundles up to 250mb

but they don't mention the actual number of bytes :sweat_smile: 

Please let me know if you think I should change

Reply:
i think it should change since we use 1024 elsewhere for arweave

Reply:
Changed in https://github.com/metaplex-foundation/metaplex/pull/859/commits/d2aa6dc93158bff811cb05fed474a875d41b0901